Add retry example for Ruby

fotos opened the pull request:
I am trying to make the Ruby server / client retries work with the native retry policy.

I have created an example server / client under `src/examples/ruby/retry` and adjusted the server to be equivalent to [`flaky_server.py` from Python](https://github.com/grpc/grpc/blob/v1.41.1/examples/python/retry/flaky_server.py) by injecting random errors. The client includes a `retryPolicy` for the `GreeterService`.

The Ruby server / client wasn't correctly retrying on `errors` and I started investigating the root cause. When I tried using the Ruby client with the Python (flaky) server the retries worked correctly. Additionally when I tried the Python client on the Ruby server it didn't work. Therefore I started digging into the Ruby server to understand what is the difference.

When running the Python example I noticed that, when a flaky response is send back, the following operations are send in one batch:

```
$ GRPC_TRACE=all GRPC_VERBOSITY=debug python ./flaky_server.py
[…]
I0910 20:42:15.883867600     538 call.cc:1573]               ops[0]: SEND_INITIAL_METADATA(nil)
I0910 20:42:15.883883900     538 call.cc:1573]               ops[1]: SEND_MESSAGE ptr=0x564ba270df40
I0910 20:42:15.883897500     538 call.cc:1573]               ops[2]: SEND_STATUS_FROM_SERVER status=0 details=(nil)
[…]
I0910 20:42:15.884151100     538 chttp2_transport.cc:1360]   HTTP:3:HDR:SVR: :status: 200
I0910 20:42:15.884159800     538 chttp2_transport.cc:1360]   HTTP:3:HDR:SVR: content-type: application/grpc
I0910 20:42:15.884167800     538 chttp2_transport.cc:1360]   HTTP:3:HDR:SVR: grpc-accept-encoding: identity,deflate,gzip
I0910 20:42:15.884235900     538 chttp2_transport.cc:1360]   HTTP:3:HDR:SVR: accept-encoding: identity,gzip
I0910 20:42:15.884310400     538 chttp2_transport.cc:1360]   HTTP:3:TRL:SVR: grpc-status: 0
I0910 20:42:15.884326100     538 chttp2_transport.cc:1360]   HTTP:3:TRL:SVR: grpc-message:
```

Both the `SEND_INITIAL_METADATA` and `SEND_STATUS_FROM_SERVER` are send in one batch.

This results in the client being able to parse the Trailing Headers and initiate a retry:

```
I0911 15:51:23.677840900    2944 tcp_posix.cc:544]           TCP:0x5644cb109200 notify_on_read
I0911 15:51:23.677862000    2944 retry_filter.cc:1392]       chand=0x5644cb205470 calld=0x5644cb2030d0 attempt=0x7f1100001540 batch_data=0x7f1100001250: got recv_initial_metadata_ready, error="No Error"
I0911 15:51:23.677882100    2944 retry_filter.cc:1419]       chand=0x5644cb205470 calld=0x5644cb2030d0 attempt=0x7f1100001540: deferring recv_initial_metadata_ready (Trailers-Only)
I0911 15:51:23.677904400    2944 retry_filter.cc:1491]       chand=0x5644cb205470 calld=0x5644cb2030d0 attempt=0x7f1100001540 batch_data=0x7f1100001250: got recv_message_ready, error="No Error"
I0911 15:51:23.677932600    2944 retry_filter.cc:1517]       chand=0x5644cb205470 calld=0x5644cb2030d0 attempt=0x7f1100001540: deferring recv_message_ready (nullptr message and recv_trailing_metadata pending)
I0911 15:51:23.678026700    2944 retry_filter.cc:1687]       chand=0x5644cb205470 calld=0x5644cb2030d0 attempt=0x7f1100001540 batch_data=0x7f1100001250: got recv_trailing_metadata_ready, error="No Error"
I0911 15:51:23.678036700    2944 retry_filter.cc:1714]       chand=0x5644cb205470 calld=0x5644cb2030d0 attempt=0x7f1100001540: call finished, status=UNAVAILABLE is_lb_drop=0
I0911 15:51:23.678057500    2944 retry_filter.cc:2565]       chand=0x5644cb205470 calld=0x5644cb2030d0: retrying failed call in 206 ms
```

On the other hand the Ruby server is sending 2 batches:

```
$ GRPC_TRACE=all GRPC_VERBOSITY=debug ruby ./flaky_server.rb
[…]
reserved=(nil))
I0911 15:47:12.533849800    2820 call.cc:1573]               ops[0]: SEND_INITIAL_METADATA(nil)
I0911 15:47:12.533865300    2820 call.cc:641]                OP[server:0x55da0efe6b20]:  SEND_INITIAL_METADATA{}
I0911 15:47:12.533877000    2820 channel_stack.cc:239]       OP[message_size:0x55da0efe6b38]:  SEND_INITIAL_METADATA{}
I0911 15:47:12.533909100    2820 channel_stack.cc:239]       OP[deadline:0x55da0efe6b50]:  SEND_INITIAL_METADATA{}
I0911 15:47:12.533924300    2820 channel_stack.cc:239]       OP[http-server:0x55da0efe6b68]:  SEND_INITIAL_METADATA{}
[…]
I0911 15:47:12.534143600    2820 chttp2_transport.cc:1360]   HTTP:1:HDR:SVR: :status: 200
I0911 15:47:12.534155900    2820 chttp2_transport.cc:1360]   HTTP:1:HDR:SVR: content-type: application/grpc
I0911 15:47:12.534188100    2820 chttp2_transport.cc:1360]   HTTP:1:HDR:SVR: grpc-accept-encoding: identity,deflate,gzip
I0911 15:47:12.534222700    2820 chttp2_transport.cc:1360]   HTTP:1:HDR:SVR: accept-encoding: identity,gzip
I0911 15:47:12.534239000    2820 stream_lists.cc:125]        0x55da0efe1c20[1][svr]: add to writable
```

Followed by:

```
I0911 15:47:12.535417000    2820 call.cc:1573]               ops[0]: SEND_STATUS_FROM_SERVER status=14 details=injected_error(nil)
[…]
I0911 15:47:12.535818100    2820 chttp2_transport.cc:1360]   HTTP:1:TRL:SVR: grpc-status: 14
I0911 15:47:12.535831500    2820 chttp2_transport.cc:1360]   HTTP:1:TRL:SVR: grpc-message: injected_error
I0911 15:47:12.535839500    2820 stream_lists.cc:125]        0x55da0efe1c20[1][svr]: add to writable
```

In this case the client is committing the retry based on the first batch and then doesn't retry the call:

```
D0911 15:54:08.129073900    2992 tcp_posix.cc:692]           DATA: 00 00 6b 01 04 00 00 00 01 88 40 0c 63 6f 6e 74 65 6e 74 2d 74 79 70 65 10 61 70 70 6c 69 63 61 74 69 6f 6e 2f 67 72 70 63 40 14 67 72 70 63 2d 61 63 63 65 70 74 2d 65 6e 63 6f 64 69 6e 67 15 69 64 65 6e 74 69 74 79 2c 64 65 66 6c 61 74 65 2c 67 7a 69 70 40 0f 61 63 63 65 70 74 2d 65 6e 63 6f 64 69 6e 67 0d 69 64 65 6e 74 69 74 79 2c 67 7a 69 70 00 00 2d 01 05 00 00 00 01 40 0b 67 72 70 63 2d 73 74 61 74 75 73 02 31 34 40 0c 67 72 70 63 2d 6d 65 73 73 61 67 65 0e 69 6e 6a 65 63 74 65 64 5f 65 72 72 6f 72 '..k.......@.content-type.application/grpc@.grpc-accept-encoding.identity,deflate,gzip@.accept-encoding.identity,gzip..-......@.grpc-status.14@.grpc-message.injected_error'
I0911 15:54:08.129004100    2987 tcp_posix.cc:1572]          WRITE 0x561d69f3ccf0 (peer=ipv4:127.0.0.1:50051)
D0911 15:54:08.129400600    2987 tcp_posix.cc:1576]          DATA: 00 00 08 06 01 00 00 00 00 00 00 00 00 00 00 00 00 '.................'
I0911 15:54:08.129764900    2987 tcp_posix.cc:1621]          write: "No Error"
I0911 15:54:08.129816800    2987 parsing.cc:686]             parsing initial_metadata
I0911 15:54:08.129841300    2987 parsing.cc:433]             HTTP:1:HDR:CLI: :status: 32 30 30 '200'
I0911 15:54:08.129880700    2987 parsing.cc:433]             HTTP:1:HDR:CLI: content-type: 61 70 70 6c 69 63 61 74 69 6f 6e 2f 67 72 70 63 'application/grpc'
I0911 15:54:08.129907000    2987 parsing.cc:433]             HTTP:1:HDR:CLI: grpc-accept-encoding: 69 64 65 6e 74 69 74 79 2c 64 65 66 6c 61 74 65 2c 67 7a 69 70 'identity,deflate,gzip'
I0911 15:54:08.130359200    2987 parsing.cc:433]             HTTP:1:HDR:CLI: accept-encoding: 69 64 65 6e 74 69 74 79 2c 67 7a 69 70 'identity,gzip'
I0911 15:54:08.130407500    2987 parsing.cc:691]             parsing trailing_metadata
I0911 15:54:08.130754400    2987 parsing.cc:545]             HTTP:1:TRL:CLI: grpc-status: 31 34 '14'
I0911 15:54:08.131115700    2987 parsing.cc:545]             HTTP:1:TRL:CLI: grpc-message: 69 6e 6a 65 63 74 65 64 5f 65 72 72 6f 72 'injected_error'
I0911 15:54:08.131152500    2987 tcp_posix.cc:544]           TCP:0x561d69f3ccf0 notify_on_read
I0911 15:54:08.131216500    2987 retry_filter.cc:1392]       chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130 batch_data=0x561d6a036680: got recv_initial_metadata_ready, error="No Error"
I0911 15:54:08.131227900    2987 retry_filter.cc:2534]       chand=0x561d6a0392b0 calld=0x561d6a036090: committing retries
I0911 15:54:08.131235600    2987 retry_filter.cc:2338]       chand=0x561d6a0392b0 calld=0x561d6a036090: destroying send_initial_metadata
I0911 15:54:08.131391200    2987 retry_filter.cc:2346]       chand=0x561d6a0392b0 calld=0x561d6a036090: destroying send_messages[0]
I0911 15:54:08.131407500    2987 retry_filter.cc:2355]       chand=0x561d6a0392b0 calld=0x561d6a036090: destroying send_trailing_metadata
I0911 15:54:08.131421100    2987 retry_filter.cc:775]        chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130: retry state no longer needed; moving LB call to parent and unreffing the call attempt
I0911 15:54:08.131518400    2987 retry_filter.cc:2516]       chand=0x561d6a0392b0 calld=0x561d6a036090: invoking recv_initial_metadata_ready for pending batch at index 0
I0911 15:54:08.131599700    2987 retry_filter.cc:1491]       chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130 batch_data=0x561d6a036680: got recv_message_ready, error="No Error"
I0911 15:54:08.131614600    2987 retry_filter.cc:2516]       chand=0x561d6a0392b0 calld=0x561d6a036090: invoking recv_message_ready for pending batch at index 0
I0911 15:54:08.131712000    2987 retry_filter.cc:1687]       chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130 batch_data=0x561d6a036680: got recv_trailing_metadata_ready, error="No Error"
I0911 15:54:08.131730600    2987 retry_filter.cc:1714]       chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130: call finished, status=UNAVAILABLE is_lb_drop=0
I0911 15:54:08.131764500    2987 retry_filter.cc:1122]       chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130: retries already committed
I0911 15:54:08.131780100    2987 retry_filter.cc:2516]       chand=0x561d6a0392b0 calld=0x561d6a036090: invoking recv_trailing_metadata_ready for pending batch at index 0
I0911 15:54:08.131793100    2987 retry_filter.cc:2460]       chand=0x561d6a0392b0 calld=0x561d6a036090: clearing pending batch
I0911 15:54:08.131826500    2987 retry_filter.cc:1307]       chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130: destroying batch 0x561d6a036680
I0911 15:54:08.131839500    2987 retry_filter.cc:709]        chand=0x561d6a0392b0 calld=0x561d6a036090 attempt=0x561d6a03a130: destroying call attempt
D0911 15:54:08.131971800    2987 call.cc:748]                set_final_status CLI
D0911 15:54:08.132008700    2987 call.cc:749]                {"created":"@1631375648.131931200","description":"Error received from peer ipv4:127.0.0.1:50051","file":"src/core/lib/surface/call.cc","file_line":1069,"grpc_message":"injected_error","grpc_status":14}
```

So I traced the Ruby server code a bit.

When an exception is raised it is [rescued](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/rpc_desc.rb#L130-L134) in [`GRPC::RpcDesc#run_server_method`](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/rpc_desc.rb#L117):

https://github.com/grpc/grpc/blob/ab6beb3f686857d687bf72e80b1fbaf7d3c3e4a4/src/ruby/lib/grpc/generic/rpc_desc.rb#L130-L134

Then [`GRPC::RpcDesc#send_status`](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/rpc_desc.rb#L195-L198):

https://github.com/grpc/grpc/blob/ab6beb3f686857d687bf72e80b1fbaf7d3c3e4a4/src/ruby/lib/grpc/generic/rpc_desc.rb#L195-L198

calls [`GRPC::ActiveCall#send_status`](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/active_call.rb#L205-L216):

https://github.com/grpc/grpc/blob/ab6beb3f686857d687bf72e80b1fbaf7d3c3e4a4/src/ruby/lib/grpc/generic/active_call.rb#L205-L216

which in turn calls [`GRPC::ActiveCall#send_initial_metadata`](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/active_call.rb#L118-L125):

https://github.com/grpc/grpc/blob/ab6beb3f686857d687bf72e80b1fbaf7d3c3e4a4/src/ruby/lib/grpc/generic/active_call.rb#L118-L125

finally resulting in a call to [`GRPC::ActiveCall.client_invoke`](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/active_call.rb#L63-L66):

https://github.com/grpc/grpc/blob/ab6beb3f686857d687bf72e80b1fbaf7d3c3e4a4/src/ruby/lib/grpc/generic/active_call.rb#L63-L66

The `.client_invoke` call sends a batch with only `SEND_INITIAL_METADATA`.

Subsequently back in [`GRPC::ActiveCall#send_status`](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/active_call.rb#L205-L216) the `SEND_STATUS_FROM_SERVER` is [sent as another batch](https://github.com/grpc/grpc/blob/v1.40.0/src/ruby/lib/grpc/generic/active_call.rb#L208-L212):

https://github.com/grpc/grpc/blob/ab6beb3f686857d687bf72e80b1fbaf7d3c3e4a4/src/ruby/lib/grpc/generic/active_call.rb#L208-L212

The above sequence explains the 2 batches seen in the server logs.

Changing `GRPC::ActiveCall#send_status`, per this diff, to send the initial metadata along with the server status, makes the client correctly retry failing responses:

```diff
diff --git a/src/ruby/lib/grpc/generic/active_call.rb b/src/ruby/lib/grpc/generic/active_call.rb
index 11db1f448c..0c19e5c8c5 100644
--- a/src/ruby/lib/grpc/generic/active_call.rb
+++ b/src/ruby/lib/grpc/generic/active_call.rb
@@ -204,8 +204,9 @@ module GRPC
     # list, mulitple metadata for its key are sent
     def send_status(code = OK, details = '', assert_finished = false,
                     metadata: {})
-      send_initial_metadata
+      #send_initial_metadata
       ops = {
+        SEND_INITIAL_METADATA => metadata,
         SEND_STATUS_FROM_SERVER => Struct::Status.new(code, details, metadata)
       }
       ops[RECV_CLOSE_ON_SERVER] = nil if assert_finished
```

The above results in a response similar to the Python server and the client correctly retries the request.

I believe this is a bug in the Ruby server yet I'm not sure about the implications of the change (e.g. the mutex `@send_initial_md_mutex` access around an object instance variable which seems unnecessary? 🤔) or the the proper way to fix the issue. That's why I'm asking for a confirmation of the issue and guidance. 😊 

@drfloob

HannahShiSFB commented:
According to the first retry commit scenario in this proposal,
https://github.com/grpc/proposal/blob/master/A6-client-retries.md#when-retries-are-valid
it is by design that retry should be committed when client received the initial metadata, and server is not expected to send it separately from the status.

So, the analysis and proposal of fix is legit, however gRPC protocol requires initial metadata to be sent only once, therefore send_status​ should check if initial metadata had been sent previously, a proper fix should be something like the following. Could you please convert your test code to a proper test case as well?

CC: @apolcyn 

    def send_status(code = OK, details = '', assert_finished = false,
                    metadata: {})
      ops = {}
      @send_initial_md_mutex.synchronize do
        ops[SEND_INITIAL_METADATA] = @metadata_to_send unless @metadata_sent
        @metadata_sent = true
      end

      ops[SEND_STATUS_FROM_SERVER] = Struct::Status.new(code, details, metadata)
      ops[RECV_CLOSE_ON_SERVER] = nil if assert_finished
      @call.run_batch(ops)
      set_output_stream_done

      nil
    end

fotos commented:
Appreciate the feedback @HannahShiSFB! 🙏

We will look into patching `#send_status` and writing a corresponding spec as soon as possible.

fotos commented:
@stanley-cheung / @HannahShiSFB 

Updated the PR with the proposed fix. The `retry_client.rb` example now works with the `flaky_server.rb`. 🙌

Added an expectation that the metadata are actually sent from `#send_status`. Additionally I added a separate test to cover the batch run operations. I think this PR is be ready for review.

cc: @apolcyn

HannahShiSFB commented:
> @stanley-cheung / @HannahShiSFB
> 
> Updated the PR with the proposed fix. The `retry_client.rb` example now works with the `flaky_server.rb`. 🙌
> 
> Added an expectation that the metadata are actually sent from `#send_status`. Additionally I added a separate test to cover the batch run operations. I think this PR is ready for review.
> 
> cc: @apolcyn

LGTM
@stanley-cheung @apolcyn

apolcyn commented:
Thanks for the work on this.

I'm actually wondering if this behavior in the PR description could be a bug in the client-side retry code, though.

@markdroth is out of the office right now, but I'd like to get his input when he's back

fotos commented:
@apolcyn 

> I'm actually wondering if this behavior in the PR description could be a bug in the client-side retry code, though.

Which client-side code you mean? Using Ruby as the client side?

I have tried the Ruby client code against the Python flaky server (from the example) and the retries worked:

>  When I tried using the Ruby client with the Python (flaky) server the retries worked correctly. Additionally when I tried the Python client on the Ruby server it didn't work. Therefore I started digging into the Ruby server to understand what is the difference.

The above pointed towards a bug in the Ruby server code (which the patch proposed by @HannahShiSFB fixes).

Let me know if anything is unclear. 😊

fotos commented:
@apolcyn / @markdroth any updates here? 🙏

dennycd commented:
@apolcyn  can you have a look and see if this one is ready to merge ? thanks

markdroth commented:
@apolcyn What do you think is a bug here?  The discussion here seems correct to me.  If initial metadata and status are sent in separate batches from the server, then the server will not send a Trailers-Only response, and the client cannot retry.  That is how the retry functionality is designed to work.

fotos commented:
I rebased the branch on top of the latest `origin/master`.

It's not clear to me what's blocking this fix from getting merged. Is there anything I can do to help move this forward? 🙏

fotos commented:
`E_TIMEOUT` – I guess there is no interest in merging this PR. :x:

If the circumstances changes let me know and I'll re-open / rebase the PR again.

y-yagi commented:
@apolcyn 
What do you think this fix? I've encountered the same issue. I think other users also encountered this #36461.
I'm wondering how we can move forward with resolving the issue.
